klayout/drc/run_drc.py

```diff
@@ -0,0 +1,709 @@
+
+
+def get_results(rule_deck, rules, lyrdb, type):
+
+    mytree = ET.parse(f"{lyrdb}_{type}_gf{arguments['--gf180mcu']}.lyrdb")
+    myroot = mytree.getroot()
+
+    violated = []
+
+    for lrule in rules:
+        # Loop on database to get the violations of required rule
+        for z in myroot[7]:
+            if f"'{lrule}'" == f"{z[1].text}":
+                violated.append(lrule)
+                break
+
+    lyrdb_clean = lyrdb.split("/")[-1]
+
+    if len(violated) > 0:
+        logging.error(
+            f"\nTotal # of DRC violations in {rule_deck}.drc is {len(violated)}. Please check {lyrdb_clean}_{type}_gf{arguments['--gf180mcu']}.lyrdb file For more details"
+        )
+        logging.info("Klayout GDS DRC Not Clean")
+        logging.info(f"Violated rules are : {violated}\n")
+    else:
+        logging.info(
+            f"\nCongratulations !!. No DRC Violations found in {lyrdb_clean} for {rule_deck}.drc rule deck with switch gf{arguments['--gf180mcu']}"
+        )
+        logging.info("Klayout GDS DRC Clean\n")
+
+
+def generate_drc_run_template(drc_dir: str, run_dir: str, run_tables_list: list = []):
+    """
+    generate_drc_run_template will generate the template file to run drc in the run_dir path.
+
+    Parameters
+    ----------
+    drc_dir : str
+        Path string to the location where the DRC files would be found to get the list of the rule tables.
+    run_dir : str
+        Absolute path string to the run location where all the run output will be generated.
+    deck_name : str, optional
+        Name of the rule deck to use for generating the template, by default ""
+    run_tables_list : list, optional
+        list of target parts of the rule deck, if empty assume all of the rule tables found, by default []
+
+    Returns
+    -------
+    str
+        Absolute path to the generated DRC file.
+    """
+    if len(run_tables_list) < 1:
+        all_tables = [
+            os.path.basename(f)
+            for f in glob.glob(os.path.join(drc_dir, "rule_decks", "*.drc"))
+            if "antenna" not in f
+            and "density" not in f
+            and "main" not in f
+            and "tail" not in f
+        ]
+        deck_name = "main"
+    elif len(run_tables_list) == 1:
+        deck_name = run_tables_list[0]
+        all_tables = ["{}.drc".format(run_tables_list[0])]
+    else:
+        all_tables = ["{}.drc".format(t) for t in run_tables_list]
+        deck_name = "main"
+
+    logging.info(
+        "## Generating template with for the following rule tables: {}".format(
+            str(all_tables)
+        )
+    )
+    print(run_dir)
+
+    all_tables.insert(0, "main.drc")
+    all_tables.append("tail.drc")
+
+    gen_rule_deck_path = os.path.join(run_dir, "{}.drc".format(deck_name))
+    with open(gen_rule_deck_path, "wb") as wfd:
+        for f in all_tables:
+            with open(os.path.join(drc_dir, "rule_decks", f), "rb") as fd:
+                shutil.copyfileobj(fd, wfd)
```

Comment:
wouldn't that already overwrite `wfd`?

Reply:
No, it concatenate to the file.

Reply:
It's a very efficient way to concatenate files in python.